clients/src/main/java/org/apache/kafka/common/utils/Bytes.java

```diff
@@ -0,0 +1,192 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.utils;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Comparator;
+
+/**
+ * Utility class that handles immutable byte arrays.
+ */
+public class Bytes implements Comparable<Bytes> {
+
+    private final byte[] bytes;
+
+    private final int offset;
+
+    private final int length;
+
+    private int hashCode;
+
+    private boolean hashCodeComputed;
+
+    public static Bytes wrap(byte[] bytes) {
+        return new Bytes(bytes);
+    }
+
+    /**
+     * Create a Bytes using the byte array.
+     *
+     * @param bytes This array becomes the backing storage for the object.
+     */
+    public Bytes(byte[] bytes) {
+        this(bytes, 0, bytes.length);
+    }
+
+    /**
+     * Create a Bytes using the byte array.
+     *
+     * @param bytes the byte array to set to
+     * @param offset the offset in {@code bytes} to start at
+     * @param length the number of bytes in {@code bytes}
+     */
+    public Bytes(final byte[] bytes, final int offset, final int length) {
```

Comment:
I would have expected that this constructor would "extract" a sub-array from the input `bytes` based on `offset` and `length.  I would not have expected us to keep track of the original `bytes` and offset and length.  Why do we need to do that?